### PR TITLE
Add new method to ArchiveWrite- add_entry_from_memory

### PR DESCRIPTION
**Pull request by krzysztofzuraw**
Add new method `add_entry_from_memory` for creating archive from files stored in memory.

### CODE REVIEW

File: libarchive/write.py [hunk 1]
```diff
@@ -73,6 +74,39 @@ def add_files(self, *paths):
                         write_finish_entry(write_p)
                         entry_clear(entry_p)
 
+    def add_entry_from_memory(self, entry):
```
Comment (Changaco) [inline libarchive/write.py]: Why an entry object instead of passing directly the 3 arguments?

File: libarchive/write.py [hunk 2]
```diff
@@ -73,6 +74,39 @@ def add_files(self, *paths):
...
+                write_data(archive_pointer, chunk, len(chunk))
+
+            self._close_entry_archive(archive_entry_pointer)
+
+    def _create_entry_archive(self, archive_entry_pointer, entry):
+        """Helper for setting up necessary entry parameters."""
+        archive_entry = ArchiveEntry(None, archive_entry_pointer)
+
+        entry_set_pathname(archive_entry_pointer, entry.path)
+        archive_entry.pathname = entry.path
```
Comment (Changaco) [inline libarchive/write.py]: Why are you setting `pathname` twice here using different functions?

File: libarchive/write.py (hunk 1 above)
Comment (krzysztofzuraw) [inline libarchive/write.py]: I like more consistency of one object with arguments than 3 arguments but I'm open to discussion and I can change that if you want.

**Comment by Changaco** [comment]
Note: #26 is a related issue.

File: libarchive/write.py (hunk 2 above)
Comment (krzysztofzuraw) [inline libarchive/write.py]: Done. I've removed unnecessary function.

File: libarchive/write.py (hunk 1 above)
Comment (Changaco) [inline libarchive/write.py]: Please change it. It's inconsistent and confusing, because the type of `entry` is `ArchiveEntry` elsewhere in the code, but here it's a "virtual" `Entry` type that the library doesn't provide (in the test you create it with `namedtuple`). That would probably be okay in JavaScript code, but this is Python.

Comment (krzysztofzuraw) [inline libarchive/write.py]: @Changaco changed, I also added `archive_obj_to_write` to ease extending libarchive with custom writers: 9bb50158cc707599f5f974ae4b22c0aa6ce78182

**Comment by Changaco** [comment]
A few more tweaks and this should be ready for merge.

File: libarchive/ffi.py [hunk 3]
```diff
@@ -22,7 +22,6 @@
 logger = logging.getLogger('libarchive')
 
 page_size = mmap.PAGESIZE
-
```
Comment (Changaco) [inline libarchive/ffi.py]: Please remove that unnecessary whitespace change.

File: libarchive/ffi.py [hunk 4]
```diff
@@ -46,6 +45,9 @@
 CLOSE_CALLBACK = CFUNCTYPE(c_int, c_void_p, c_void_p)
 VOID_CB = lambda *_: ARCHIVE_OK
 
+# Other types
+REGULAR_FILE = 0o100000
+DEFAULT_UNIX_PERMISSION = 0o664
```
Comment (Changaco) [inline libarchive/ffi.py]: This should be in the "Constants" section above.

File: libarchive/write.py [hunk 5]
```diff
@@ -73,6 +73,47 @@ def add_files(self, *paths):
                         write_finish_entry(write_p)
                         entry_clear(entry_p)
 
+    def add_entry_from_memory(self, entry_path, entry_size, entry_data):
```
Comment (Changaco) [inline libarchive/write.py]: Please rename this method to `add_file_from_memory`, inline the `_create_entry_archive` and `_close_entry_archive` methods into it, and add `filetype=REGULAR_FILE, permission=DEFAULT_UNIX_PERMISSION` to the arguments.

File: libarchive/write.py [hunk 6]
```diff
@@ -103,27 +147,34 @@ def write_cb_internal(archive_p, context, buffer_, length):
         ffi.write_set_bytes_in_last_block(archive_p, 1)
         ffi.write_set_bytes_per_block(archive_p, block_size)
         ffi.write_open(archive_p, None, open_cb, write_cb, close_cb)
-        yield ArchiveWrite(archive_p)
+        yield archive_obj_to_write(archive_p)
```
Comment (Changaco) [inline libarchive/write.py]: `archive_obj_to_write` is a misleading name, please change it to something like `ArchiveWriter` or `ArchiveWriteClass` instead.

File: libarchive/write.py (hunk 5 above)
Comment (krzysztofzuraw) [inline libarchive/write.py]: @Changaco changes made, but I want to separate `_create_entry_archive` and `_close_entry_archive`. In my own code I subclass `ArchiveWrite` and rewrite body in `add_file_from_memory` without need for changing `_create_entry_archive` and `_close_entry_archive`.

Comment (Changaco) [inline libarchive/write.py]: @krzysztofzuraw Building your subclass on "private" methods seems like a bad idea to me. It's fragile. Someone (probably me) is going to drop or modify these methods someday and it'll break your code.

Comment (krzysztofzuraw) [inline libarchive/write.py]: @Changaco: changed- I removed these private methods and moved code to `add_file_from_memory`

**Comment by Changaco** [comment]
Thanks @krzysztofzuraw, good work. :-)
